File: src/main/java/com/burgstaller/okhttp/DispatchingAuthenticator.java
package com.burgstaller.okhttp;

import com.burgstaller.okhttp.digest.CachingAuthenticator;

import okhttp3.Authenticator;
import okhttp3.Challenge;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.Route;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A dispatching authenticator which can be used with multiple auth schemes.
 */
public class DispatchingAuthenticator implements CachingAuthenticator {
    private final Map<String, Authenticator> authenticatorRegistry;
    private final Map<String, CachingAuthenticator> cachingRegistry;

    public DispatchingAuthenticator(Map<String, Authenticator> registry) {
        authenticatorRegistry = registry;
        cachingRegistry = new HashMap<>();
        for (Map.Entry<String, Authenticator> entry : authenticatorRegistry.entrySet()) {
            if (entry.getValue() instanceof CachingAuthenticator) {
                cachingRegistry.put(entry.getKey(), (CachingAuthenticator) entry.getValue());
            }
        }
    }

    @Override
    public Request authenticate(Route route, Response response) throws IOException {
        List<Challenge> challenges = response.challenges();
        if (!challenges.isEmpty()) {
            for (Challenge challenge : challenges) {
                final String scheme = challenge.scheme();
                Authenticator authenticator = null;
                if (scheme != null) {
                    authenticator = authenticatorRegistry.get(scheme.toLowerCase());
                }
                if (authenticator != null) {
                    return authenticator.authenticate(route, response);
                }
            }
        }
        throw new IllegalArgumentException("unsupported auth scheme " + challenges);
    }

    @Override
    public Request authenticateWithState(Request request) throws IOException {
        for (Map.Entry<String, CachingAuthenticator> authenticatorEntry : cachingRegistry.entrySet()) {
            final Request authRequest = authenticatorEntry.getValue().authenticateWithState(request);
            if (authRequest != null) {
                return authRequest;
            }
        }
        return null;
    }

    public static final class Builder {
        Map<String, Authenticator> registry = new HashMap<>();

        public Builder with(String scheme, Authenticator authenticator) {
            registry.put(scheme, authenticator);
            return this;
        }

        public DispatchingAuthenticator build() {
            return new DispatchingAuthenticator(registry);
        }
    }
}

File: src/main/kotlin/com/burgstaller/okhttp/DispatchingAuthenticator.kt
package com.burgstaller.okhttp

import com.burgstaller.okhttp.digest.CachingAuthenticator
import okhttp3.Authenticator
import okhttp3.Request
import okhttp3.Response
import okhttp3.Route
import java.util.LinkedHashMap
import java.util.Locale

/**
 * A dispatching authenticator which can be used with multiple auth schemes.
 */
class DispatchingAuthenticator private constructor(private val authenticatorRegistry: Map<String, Authenticator>) : CachingAuthenticator {
  private val cachingRegistry: MutableMap<String, CachingAuthenticator>

  init {
    cachingRegistry = LinkedHashMap()
    for ((key, value) in authenticatorRegistry) {
      if (value is CachingAuthenticator) {
        cachingRegistry[key] = value
      }
    }
  }

  override fun authenticate(route: Route?, response: Response): Request? {
    val challenges = response.challenges()
    if (!challenges.isEmpty()) {
      for (challenge in challenges) {
        val scheme = challenge.scheme()
        var authenticator: Authenticator? = null
        if (scheme != null) {
          authenticator = authenticatorRegistry[scheme.toLowerCase(Locale.getDefault())]
        }
        if (authenticator != null) {
          return authenticator.authenticate(route, response)
        }
      }
    }
    throw IllegalArgumentException("unsupported auth scheme $challenges")
  }

  override fun authenticateWithState(route: Route?, request: Request): Request? {
    for ((_, value) in cachingRegistry) {
      val authRequest = value.authenticateWithState(route, request)
      if (authRequest != null) {
        return authRequest
      }
    }
    return null
  }

  class Builder {
    private var registry: MutableMap<String, Authenticator> = LinkedHashMap()

    fun with(scheme: String, authenticator: Authenticator): Builder {
      registry[scheme.toLowerCase(Locale.getDefault())] = authenticator
      return this
    }

    fun build(): DispatchingAuthenticator {
      return DispatchingAuthenticator(registry)
    }
  }
}

